### PR TITLE
HPCC4J-611 Add OpenTelemetry tracing to dfsclient

### PR DESCRIPTION
**Pull request by jpmcmu**
- Updated event attribute reporting
- Fixed trace structure in dfsclient protocol
- Added tracing to FileUtility

Signed-off-by: James McMullan James.McMullan@lexisnexis.com

## Type of change:
- [X] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [ ] This change does not cause any existing JUnits to fail.
  - [ ] I have include JUnit coverage to test this change
  - [ ] I have performed system test and covered possible regressions and side effects.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] This change fixes the problem, not just the symptom

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana Wrapped up code review changes, please review

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java [hunk 1]
```diff
@@ -129,6 +161,15 @@ public boolean hasOperation()
 
         public void startOperation(String operationName)
         {
+            if (taskSpan != null)
+            {
+                taskSpan = Utils.createChildSpan(taskSpan, operationName);
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java:166]: does it makes sense to have a single Utils.createSpan(parentSpan, spanName)? If parentSpan != null, it sets it as the parent

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java [hunk 2]
```diff
@@ -146,6 +187,17 @@ public void endOperation(boolean success)
...
+            }
+            else
+            {
+                taskSpan.setStatus(StatusCode.ERROR);
+            }
+
+            taskSpan.end();
+
             long totalOperationTime = System.nanoTime();
             totalOperationTime -= operationStart;
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java:203]: some of these context values seem worth reporting in the span

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java [hunk 3]
```diff
@@ -1095,7 +1149,8 @@ private static void performRead(String[] args, TaskContext context)
         for (int i = 0; i < datasets.length; i++)
         {
             String datasetName = datasets[i];
-            context.startOperation("Read " + datasetName);
+            context.startOperation("FileUtility.Read_" + datasetName);
+            context.setTaskSpanAttributes(Attributes.of(AttributeKey.stringKey("server.url"), connString));
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java:1153]: is there a way to add individual attributes after this initial set attributes call?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java [hunk 4]
```diff
@@ -1279,7 +1336,9 @@ private static void performReadTest(String[] args, TaskContext context)
         }
 
         String datasetName = cmd.getOptionValue("read_test");
-        context.startOperation("Read Test " + datasetName);
+        context.startOperation("FileUtility.ReadTest_" + datasetName);
+
+        context.setTaskSpanAttributes(Attributes.of(AttributeKey.stringKey("server.url"), connString));
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java:1341]: Are there other attributes we might want to report to help track issues?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java [hunk 5]
```diff
@@ -1463,7 +1521,9 @@ private static void performCopy(String[] args, TaskContext context)
             String srcFile = copyPairs[i];
             String destFile = copyPairs[i+1];
 
-            context.startOperation("Copy " + srcFile + " -> " + destFile);
+            context.startOperation("FileUtility.Copy_ " + srcFile + " -> " + destFile);
+            context.setTaskSpanAttributes(Attributes.of(AttributeKey.stringKey("server.src.url"), srcURL,
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java:1525]: does the context.staroperation call add any relevant attributes to the span?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/HPCCRemoteFileWriter.java [hunk 6]
```diff
@@ -36,17 +36,38 @@ public class HPCCRemoteFileWriter<T>
...
+        public FieldDef recordDef = null;
+        public CompressionAlgorithm fileCompression = CompressionAlgorithm.DEFAULT;
+        public int connectTimeoutMs = -1;
+        public int socketOpTimeoutMs = -1;
+        public Span parentSpan = null;
+    }
+
+    private static FileWriteContext constructReadContext(FieldDef recordDef, CompressionAlgorithm fileCompression, int connectTimeoutMs, int socketOpTimeoutMs)
+    {
+        FileWriteContext context = new FileWriteContext();
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/HPCCRemoteFileWriter.java:62]: do all/some of these end up tracked in a span?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/Utils.java [hunk 7]
```diff
@@ -13,31 +13,60 @@
...
 public class Utils
 {
+    private static OpenTelemetry globalOpenTelemetry = null;
+    private static Tracer dfsClientTracer = null;
+
+    public static OpenTelemetry getOpenTelemetry()
+    {
+        if (globalOpenTelemetry == null)
+        {
+            globalOpenTelemetry = GlobalOpenTelemetry.get();
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/Utils.java:31]: should this call setGlobalOpenTelemetry ?

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/Utils.java [hunk 8]
```diff
@@ -13,31 +13,60 @@
...
+            dfsClientTracer = getOpenTelemetry().getTracer("org.hpccsystems.dfsclient");
+        }
+
+        return dfsClientTracer;
     }
 
     public static Span createSpan(String name)
     {
-        return Utils.getTelemetryTracer().spanBuilder(name)
+        Span span = Utils.getTelemetryTracer().spanBuilder(name)
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/Utils.java:54]: can we combine the two functions?
Span createSpan(String name, Span parent)

**Review by rpastrana — CHANGES_REQUESTED** [review]
@jpmcmu looks good, but left you a few suggestions. Let's discuss.

**Comment by jpmcmu** [comment]
@rpastrana please review

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/Utils.java (hunk 8 above)
Comment (rpastrana) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/Utils.java:49]: are we combining these 2?

**Review by rpastrana — CHANGES_REQUESTED** [review]
@jpmcmu reviwed

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java [hunk 9]
```diff
@@ -66,72 +74,206 @@ public class FileUtility
...
-            return !currentOperationDesc.isEmpty();
+            if (operations.empty())
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        public TaskOperation getOperation()
```
[the remark this answers is not in the document]
Comment (jpmcmu) [reply dfsclient/src/main/java/org/hpccsystems/dfs/client/FileUtility.java:251]: I want all new operations to go through startOperation / endOperation. So I don't want to expose an addOperation etc

**Review by rpastrana — APPROVED** [review]

**Comment by jpmcmu** [comment]
Tracing results from Kibana:
Screenshot 2024-07-19 at 2 22 27 PM
Screenshot 2024-07-19 at 2 22 50 PM

**Comment by jpmcmu** [comment]
With ESP tracing enabled:
Screenshot 2024-07-19 at 4 17 02 PM

**Comment by rpastrana** [comment]
> Tracing results from Kibana, second screenshot is the list of events under the read span. Screenshot 2024-07-19 at 2 22 27 PM Screenshot 2024-07-19 at 2 22 50 PM

@jakesmith once we get dafilesrv instrumentation we'll be able to track these transactions end-to-end. We're able to deploy small, local Elastic stack based trace aggregator which would be very helpful in the instrumentation process. We can share that info if interested.

**Pull request merged**